omegaconf/dictconfig.py

```diff
@@ -146,103 +145,103 @@ def _validate_get(self, key: Any, value: Any = None) -> None:
                     key=key, value=value, cause=ConfigAttributeError(msg)
                 )
 
-    def _validate_merge(self, key: Any, value: Any) -> None:
-        self._validate_set_merge_impl(key, value, is_assign=False)
-
     def _validate_set(self, key: Any, value: Any) -> None:
-        self._validate_set_merge_impl(key, value, is_assign=True)
-
-    def _validate_set_merge_impl(self, key: Any, value: Any, is_assign: bool) -> None:
         from omegaconf import OmegaConf
 
         vk = get_value_kind(value)
         if vk in (ValueKind.INTERPOLATION, ValueKind.STR_INTERPOLATION):
             return
-
-        if OmegaConf.is_none(value):
-            if key is not None:
-                child = self._get_node(key)
-                if child is not None and not child._is_optional():
-                    self._format_and_raise(
-                        key=key,
-                        value=value,
-                        cause=ValidationError("child '$FULL_KEY' is not Optional"),
-                    )
-            else:
-                if not self._is_optional():
-                    self._format_and_raise(
-                        key=None,
-                        value=value,
-                        cause=ValidationError("field '$FULL_KEY' is not Optional"),
-                    )
-
-        if value == "???":
+        self._validate_non_optional(key, value)
+        if value == "???" or value is None:
             return
 
-        if is_assign and isinstance(value, ValueNode) and self._has_element_type():
-            self._check_assign_value_node(key, value)
-            return
+        target = self._get_node(key) if key is not None else self
 
-        target: Optional[Node]
-        if key is None:
-            target = self
-        else:
-            target = self._get_node(key)
+        assign_value_node = isinstance(value, ValueNode) and self._has_element_type()
+        target_has_ref_type = isinstance(
+            target, DictConfig
+        ) and target._metadata.ref_type not in (Any, dict)
+        is_invalid_target = target is None or not target_has_ref_type
 
-        if target is None:
+        if is_invalid_target and not assign_value_node:
```

Comment:
This reads wrong, why would we want to return on an invalid target? (whatever that means)?